app/main.py

```diff
@@ -87,127 +1,54 @@
+            param["power"],
+            param["hp"]
+        )
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+    ## АDD KNIGHTS EQUIPMENTS
+    for knight_name in all_knights:
+        config = knights_config[knight_name]
+        knights = all_knights[knight_name]
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+        if "armour" in config:
+            armour_list = config["armour"]
+            knights.add_armours(armour_list)
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
+        if "potion" in config and type(config["potion"]) == dict:
+            knights.add_potion(config["potion"])
 
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+        knights.add_weapon(config["weapon"])
 
     # -------------------------------------------------------------------------------
     # BATTLE:
+    ## tournament participants
+    lancelot = all_knights["lancelot"]
+    mordred = all_knights["mordred"]
+    arthur = all_knights["arthur"]
+    red_knight = all_knights["red_knight"]
 
     # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
+    lancelot.attack(mordred)
+    mordred.attack(lancelot)
     # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+    arthur.attack(red_knight)
+    red_knight.attack(arthur)
```

Comment:
Similar to the previous comment, ensure that a knight only attacks if they are still 'alive' (hp > 0) after the first attack in the battle sequence.